File: Source/Utility/Hooks/use-motion.hook.ts
import Ripple from "@rbxts/ripple";
import Roact from "@rbxts/roact";

import { useBinding, useMemo } from "@rbxts/roact-hooked";
import { useJanitor } from "./use-once.hook";

export function useMotion(goal: number): LuaTuple<[ Roact.Binding<number>, Ripple.Motion<number>]>;

export function useMotion<T extends Ripple.MotionGoal>(goal: T): LuaTuple<[ Roact.Binding<T>, Ripple.Motion<T>]>;

export function useMotion<T extends Ripple.MotionGoal>(goal: T) {
	const motion = useMemo(() => {
		return Ripple.createMotion(goal, { start: true });
	}, []);

	const [binding, setValue] = useBinding(motion.get());

	useJanitor(Garbage => {
		Garbage.Add(motion.onStep(setValue), true);
		Garbage.Add(motion, "destroy");
	});

	return $tuple(binding, motion);
};

File: Output/Utility/Hooks/use-motion.hook.lua
-- Compiled with roblox-ts v2.2.0
local TS = _G[script]
local Ripple = TS.import(script, TS.getModule(script, "@rbxts", "ripple"))
local _roact_hooked = TS.import(script, TS.getModule(script, "@rbxts", "roact-hooked").src)
local useBinding = _roact_hooked.useBinding
local useMemo = _roact_hooked.useMemo
local useJanitor = TS.import(script, script.Parent, "use-once.hook").useJanitor
local function useMotion(goal)
	local motion = useMemo(function()
		return Ripple.createMotion(goal, {
			start = true,
		})
	end, {})
	local binding, setValue = useBinding(motion:get())
	useJanitor(function(Garbage)
		Garbage:Add(motion:onStep(setValue), true)
		Garbage:Add(motion, "destroy")
	end)
	return binding, motion
end
local default = useMotion
return {
	useMotion = useMotion,
	default = default,
}
